hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/LogicalToPhysicalOpConverter.java

```diff
@@ -24,125 +13,32 @@
-		else throw new UnsupportedOperationException("Unsupported type of logical operator: " + lop.getClass().getName() + ".");
-	}
-
-	public static UnaryPhysicalOp convert( final LogicalOpGPAdd lop ) {
-		final FederationMember fm = lop.getFederationMember();
-
-		if (      fm instanceof SPARQLEndpoint ) {
-			return new PhysicalOpBindJoinWithBoundJoin(lop);
-		}
-		else if ( fm instanceof TPFServer && lop.containsTriplePatternOnly() ) {
-			return new PhysicalOpIndexNestedLoopsJoin(lop);
-		}
-		else if ( fm instanceof BRTPFServer && lop.containsTriplePatternOnly() ) {
-			return new PhysicalOpIndexNestedLoopsJoin(lop);
-		}
-
-		else throw new UnsupportedOperationException("Unsupported type of federation member: " + fm.getClass().getName() + ".");
-	}
-
-	public static UnaryPhysicalOp convert( final LogicalOpGPOptAdd lop ) {
-		final FederationMember fm = lop.getFederationMember();
-
-		if (      fm instanceof SPARQLEndpoint ) {
-			return new PhysicalOpBindJoinWithBoundJoin(lop);
-		}
-		else if ( fm instanceof TPFServer && lop.containsTriplePatternOnly() ) {
-			return new PhysicalOpIndexNestedLoopsJoin(lop);
-		}
-		else if ( fm instanceof BRTPFServer && lop.containsTriplePatternOnly() ) {
-			return new PhysicalOpIndexNestedLoopsJoin(lop);
-		}
-
-		else throw new UnsupportedOperationException("Unsupported type of federation member: " + fm.getClass().getName() + ".");
-	}
-
-	public static UnaryPhysicalOp convert( final LogicalOpFilter lop ) {
-		return new PhysicalOpFilter(lop);
-	}
-
-	public static UnaryPhysicalOp convert( final LogicalOpBind lop ) {
-		return new PhysicalOpBind(lop);
-	}
-
-	public static UnaryPhysicalOp convert( final LogicalOpLocalToGlobal lop ) {
-		return new PhysicalOpLocalToGlobal(lop);
-	}
-
-	public static UnaryPhysicalOp convert( final LogicalOpGlobalToLocal lop ) {
-		return new PhysicalOpGlobalToLocal(lop);
-	}
-
-	// --------- binary operators -----------
+	final private static PhysicalOpRegistry registry = new PhysicalOpRegistry()
+		.register( new PhysicalOpBinaryUnion.Factory() )           // Binary union
+		.register( new PhysicalOpMultiwayUnion.Factory() )         // Multiway union
+		.register( new PhysicalOpBind.Factory() )                  // Bind clauses
+		.register( new PhysicalOpFilter.Factory() )                // Filter clauses
+		.register( new PhysicalOpRequest.Factory() )               // Request at a federation member
+		.register( new PhysicalOpGlobalToLocal.Factory() )         // Apply vocab mappings global to local
+		.register( new PhysicalOpLocalToGlobal.Factory() )         // Apply vocab mappings local to global
+		.register( new PhysicalOpBindJoin.Factory() )              // Bind-join for brTPF interface
+		.register( new PhysicalOpBindJoinWithBoundJoin.Factory() ) // Bind-join for SPARQL interface
+		.register( new PhysicalOpBindJoinWithUNION.Factory() )     // (fallback) if no non-joining var available
+		// .register( new PhysicalOpBindJoinWithVALUESorFILTER.Factory() )
```

Comment:
PhysicalOpBindJoinWithVALUESorFILTER is a better fallback (at least for now -- let's see what Adriana's experiments will show us).